---
title: "repurposingFunctions"
author: "Lizzy Ramsey"
date: "4/7/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### DESeq2 Functions

Salmon merged gene counts to dds results
```{r}
salmon2dds <- function(counts, metadata) {
  #count matrix in the sumarized experiment object
  cts <- counts@assays@data@listData[["counts"]]
  #round, salmon outputs have decimals
  cts <- round(cts)
  #colnames(cts) <- str_sub(colnames(cts), 1, -2) #for GSE69556
  #remove version numbers from transcript id's
  rownames(cts) <- gsub("\\..*", "", rownames(cts))
  
  #remove all non-alphanumeric characters from Genotype
  metadata$Genotype <- str_replace_all(metadata$Genotype, "[^[:alnum:]]", "")
  #contrasts need to be factor levels
  metadata$Genotype <- as.factor(metadata$Genotype)
#print(rownames(metadata))
#print(colnames(cts))
  dds <- DESeqDataSetFromMatrix(cts, metadata, design = ~ Genotype)
return(dds)
}

```


Up and down log2fc
```{r}
uplfc <- function(res){
  #NAs removed first
  res <- res[!is.na(res$log2FoldChange),]
  #p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
  upfc <- res[res$log2FoldChange > 2.0,]

  #pull just logfold change values from DESeqResults object and name the values by associated gene
  up <- upfc$log2FoldChange
  #names(up) <- rownames(upfc)
  up <- cbind(Mouse.gene.stable.ID = rownames(upfc), UpLFC = up)
  
return(as.data.frame(up))
}

downlfc <- function(res){
  #NAs removed first
  res <- res[!is.na(res$log2FoldChange),]
  #p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
  downfc <- res[res$log2FoldChange < -2.0,]

  #pull just logfold change values from DESeqResults object and name the values by associated gene
  down <- downfc$log2FoldChange
  #names(down) <- rownames(downfc)
  down <- cbind(Mouse.gene.stable.ID = rownames(downfc), DownLFC = down)
  
return(as.data.frame(down))
}
```

### Ortholog Mapping, LINCS Filtering
Adapted code from here https://www.r-bloggers.com/2016/10/converting-mouse-to-human-gene-names-with-biomart-package/ 
```{r}
# Basic function to convert mouse to human gene names
DEPRconvertMouseGeneList <- function(x){

require("biomaRt")
human <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
mouse <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

#attributes = attributes to retrieve
genesV2 <- getLDS(attributes = c("ensembl_gene_id"), filters = "ensembl_gene_id", values = x, mart = mouse, attributesL = c("entrezgene_id"), martL = human, uniqueRows=T)
#humanx <- unique(genesV2[, 2])] distinct(.data, ..., .keep_all = FALSE) 
genesV2 <- as.data.frame(genesV2)
#filter for unique output genes
colnames(genesV2) <- c("MouseEns", "HumanEntrez")
#human <- dplyr::distinct(genesV2, HumanEntrez) 

return(genesV2)
}
```

Manual annotation,mapping
```{r}
#Ensembl mouse, human
ens_annot <- read.csv("/Users/eramsey/Downloads/PKD/kidney_drugrepurposing/data/annot_ens_humanmouse.csv")

ens_mouse2human <- merge(musdegs_19_UP, ens_annot, by = "Mouse.gene.stable.ID")

human <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
genes <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id"), values = ens_mouse2human$Gene.stable.ID, mart = human)

names(ens_mouse2human) <- c("mouse_ensembl_gene_id", "LFC", "ensembl_gene_id")
conv_tbl <- merge(genes, ens_mouse2human, by = "ensembl_gene_id")

lincs_genes <- rhdf5:::h5read(lincs, "rownames", drop=TRUE)
lincs_genes <- cbind(entrezgene_id = lincs_genes)
conv_tbl <- merge(conv_tbl, lincs_genes, by = "entrezgene_id")
conv_tbl$LFC <- as.numeric(conv_tbl$LFC)
top100 <- slice_max(conv_tbl, n = 100, order_by = LFC)
```

NEW mouse ens to human entrez, whole conversion table returned with LFC values
```{r}
convertMouseGeneList <- function(lfc){
  #read in mose to human ens annotation
  ens_annot <- read.csv("/Users/eramsey/Downloads/PKD/kidney_drugrepurposing/data/annot_ens_humanmouse.csv")
  #conversion table
  ens_mouse2human <- merge(lfc, ens_annot, by = "Mouse.gene.stable.ID")
  #biomart to conver human ens to human entrez
  human <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
  genes <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id"), values = ens_mouse2human$Gene.stable.ID, mart = human)
  
  names(ens_mouse2human) <- c("mouse_ensembl_gene_id", "LFC", "ensembl_gene_id")
  conv_tbl <- merge(genes, ens_mouse2human, by = "ensembl_gene_id")

return(conv_tbl)
}
```

NEW Filtering for genes in refdb + convertMouseGeneList, top 100 by LFC values
```{r}
##READ this in at the beginning of script
lincs_genes <- rhdf5:::h5read(lincs, "rownames", drop=TRUE)

toplincs <- function(conv_tbl, direction){
  lincs_genes <- cbind(entrezgene_id = lincs_genes)
  conv_tbl <- merge(conv_tbl, lincs_genes, by = "entrezgene_id")
  conv_tbl$LFC <- as.numeric(conv_tbl$LFC)
  if(direction == "up"){
  top100 <- slice_max(conv_tbl, n = 100, order_by = LFC)
  }
  else{
    top100 <- slice_min(conv_tbl, n = 100, order_by = LFC)
  }
return(top100)
}
```

```{r}
lincsGenes <- function(lfc, direction){
  conv_tbl <- convertMouseGeneList(lfc)
  top100 <- toplincs(conv_tbl, direction = direction)
  return(top100)
}
```

Filtering for genes in refdb + convertMouseGeneList
```{r}
DEPRlincsGenes <- function(x, direction) {
  if(direction == "up"){
    x <- sort(x, decreasing = TRUE)
    if(length(x) > 400){
      x <- x[1:400]
    }
  }
  else{
    x <- sort(x)
      if(length(x) > 400){
      x <- x[1:400]
      }
  }
  homgenes <- convertMouseGeneList(names(x))
  #make sure lincs has already been loaded
  gid_db <- rhdf5:::h5read(lincs, "rownames", drop=TRUE)
  geneset <- homgenes$HumanEntrez[homgenes$HumanEntrez %in% gid_db]
return(geneset)
}


```


```{r}


```

### SignatureSearch Functions
QuickRun signatureSearch function  
Input: upgenes and downgenes are df outputs of convertMouseGenes, and select for Entrez column (i.e. degs_UP$HumanEntrez). For cells, select "kidney" or "all" -- kidney will subset for the kidney cell lines HA1E and NKDBA
```{r}
ssRun <- function(upgenes, downgenes, cells){
  q <- qSig(query = list(upset=as.character(upgenes), downset=as.character(downgenes)), gess_method="LINCS", refdb= lincs)
  gess <- gess_lincs(q, sortby="WTCS", tau=FALSE, workers=1)
  gess_res <- result(gess)
  gess_res <- dplyr::filter(gess_res, WTCS < 0)
  if(cells == "kidney"){
    gess_kidney <- gess_res %>% dplyr::filter(cell == "HA1E" | cell == "NKDBA")
    return(gess_kidney)
  }
  else{
return(gess_res)
  }
}

```
